src/ring_theory/ore_localization/basic.lean

```diff
@@ -0,0 +1,785 @@
+import ring_theory.non_zero_divisors
+import algebra.group_with_zero.basic
+import tactic.noncomm_ring
+
+/-!
+
+# Localization over right Ore sets.
+
+This file defines the localization of a monoid over a right Ore set and proves its universal
+mapping property. It then extends the definition and its properties first to semirings and then
+to rings. We show that in the case of a commutative monoid this definition coincides with the
+common monoid localization. Finally we show that in a ring without zero divisors, taking the Ore
+localization at `R - {0}` results in a division ring.
+
+## Notations
+
+Introduces the notation `R[S⁻¹]` for the Ore localization of a monoid `R` at a right Ore
+subset `S`. Also defines a new heterogeneos division notation `r /ₒ s` for a numerator `r : R` and
+a denominator `s : S`.
+
+## References
+
+* <https://ncatlab.org/nlab/show/Ore+localization>
+* [Zoran Škoda, *Noncommutative localization in noncommutative geometry*][skoda2006]
+
+
+## Tags
+localization, Ore, non-commutative
+
+-/
+
+universe u
+
+open ore_localization
+
+namespace ore_localization
+
+variables (R : Type*) [monoid R] (S : submonoid R) [ore : ore_set S]
+
+include ore
+
+/-- The setoid on `R × S` used for the Ore localization. -/
+def ore_eqv : setoid (R × S) :=
+{ r := λ rs rs', ∃ (u : S) (v : R), rs'.1 * u = rs.1 * v
+                            ∧ (rs'.2 : R) * u = rs.2 * v,
+  iseqv :=
+  begin
+    refine ⟨_, _, _⟩,
+    { rintro ⟨r,s⟩, use 1, use 1, simp [submonoid.one_mem] },
+    { rintros ⟨r, s⟩ ⟨r', s'⟩ ⟨u, v, hru, hsu⟩,
+      rcases ore_condition (s : R) s' with ⟨r₂, s₂, h₁⟩,
+      rcases ore_condition r₂ u with ⟨r₃, s₃, h₂⟩,
+      have : (s : R) * ((v : R) * r₃) = (s : R) * (s₂ * s₃),
+      { assoc_rw [h₁, h₂, hsu], symmetry, apply mul_assoc },
+      rcases ore_left_cancel (v * r₃) (s₂ * s₃) s this with ⟨w, hw⟩,
+      use s₂ * s₃ * w, use u * r₃ * w, split; simp only [submonoid.coe_mul],
+      { assoc_rw [hru, ←hw], simp [mul_assoc] },
+      { assoc_rw [hsu, ←hw], simp [mul_assoc] } },
+    { rintros ⟨r₁, s₁⟩ ⟨r₂, s₂⟩ ⟨r₃, s₃⟩ ⟨u, v, hur₁, hs₁u⟩ ⟨u', v', hur₂, hs₂u⟩,
+      rcases ore_condition v' u with ⟨r', s', h⟩,
+      use u' * s', use v * r', split; simp only [submonoid.coe_mul],
+      { assoc_rw [hur₂, h, hur₁, mul_assoc] },
+      { assoc_rw [hs₂u, h, hs₁u, mul_assoc] } }
+  end }
+
+end ore_localization
+
+/-- The ore localization of a monoid and a submonoid fulfilling the ore condition. -/
+def ore_localization (R : Type*) [monoid R] (S : submonoid R) [ore : ore_set S] :=
+quotient (ore_localization.ore_eqv R S)
+
+namespace ore_localization
+
+section monoid
+
+variables {R : Type*} [monoid R] {S : submonoid R}
+
+/-- Helper lemma to convert between two common forms of equality in `R` and `S`, used in a lot
+of constructions in this file. -/
+lemma submonoid.eq_of_coe_eq_mul {s s' : S} {r : R} (h : (s : R) = s' * r) :
+  s = ⟨s' * r, by { rw [←h], simp }⟩ :=
+by { ext, simp [h] }
```

Comment:
Hmm, I usually need `submonoid.eq_of_coe_eq_mul` in exactly this form, with the multiplication on on side, so that's not generalisable to `subtype`...